src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,416 @@
+        -- (7) the concatenation of all keys
+        <> mconcat [ BB.shortByteString (SBS ba#)
+                   | S.SerialisedKey (ByteArray ba#) <- toList pageKeys
+                   ]
+        -- (8) the concatenation of all values
+        <> mconcat [ BB.shortByteString (SBS ba#) -- TODO: wrong
+                   | RawValue _ _ (ByteArray ba#) <- toList pageValues
+                   ]
+  where
+    PageSize{..} = pageSize
+
+    n = pageSizeNumElems
+    b = pageSizeNumBlobs
+
+    -- Offset to component (5)
+    offKeyOffsets =
+         8                                         -- size of (1)
+      + (n + 63) `shiftR` 3 .&. complement 0x7     -- size of (2)
+      + (2 * n + 63) `shiftR` 3 .&. complement 0x7 -- size of (3)
+      + (4 + 8) * b                                -- size of (4)
+
+    -- Offset to component (7)
+    offKeys =
+        offKeyOffsets
+      + 2 * n                             -- size of (5)
+      + (if n == 1 then 6 else 2 * (n+1)) -- size of (6)
+
+    -- Thes values start right after the keys,
+    (pageKeyOffsets, offValues) =
+        case Seq.scanl (\o k -> o + sizeofKey16 k) offKeys pageKeys of
+          Empty      -> error "impossible"
+          kos :|> vo -> (kos, vo)
+
+    pageValueOffsets = case pageValues of
+      v :<| Empty -> Right (offValues, fromIntegral offValues + sizeofValue32 v)
+      _   -> Left (Seq.scanl (\o v -> o + sizeofValue16 v) offValues pageValues)
+
+{-------------------------------------------------------------------------------
+  Intermediate representation for a page
+-------------------------------------------------------------------------------}
+
+-- We might have preferred a representation for an incrementally constructed
+-- page that already builds up the raw bytes of a page. However, we do not know
+-- a priori what the offsets for most of the page components will be. These
+-- offsets depend on the number of keys\/values\/blob references, and the sizes
+-- of keys\/values, and we only find out these numbers as key-operation pairs
+-- are added incrementally. As such, we can only construct the raw page bytes
+-- once the current page has filled up.
+data PageIntermediate = PageIntermediate {
+    -- | (1) directory of components
+    pageSize           :: !PageSize
+    -- | (2) an array of 1-bit blob reference indicators
+  , pageBlobRefBitmap  :: !BitMap
+    -- | (3) an array of 2-bit operation types
+  , pageOperations     :: !CrumbMap
+    -- | (4) a pair of arrays of blob references
+  , pageBlobRefOffsets :: !(Seq Word64)
+    -- | (4) a pair of arrays of blob references, ctd
+  , pageBlobRefSizes   :: !(Seq Word32)
+    --   (5) key offsets will be computed when serialising the page
+    --   (6) value offsets will be computed when serialising the page
+    -- | (7) the concatenation of all keys
+  , pageKeys           :: !(Seq RawKey)
+    -- | (8) the concatenation of all values
+  , pageValues         :: !(Seq RawValue)
+  }
```

Comment:
An alternative approach to `Seq` would be to use lists, which have faster folding behaviour, which could be useful for serialisation. However, the disadvantages of list outweigh that advantage:
* Lists aren't strict
* Lists are accumulated in reverse, and we have to reverse them before serialisation

Reply:
- `Seq` isn't strict either, Seq's spine specifically isn't strict.
- Folding the list from the right isn't that bad (O(n) stack space), alternatevely a specialized list which tracks offsets, so the list can be traversed from the left, but elements would be written in reverse.

Both arguments are micro-optimizations. But I think that for small-count pages, the `Seq`s terrible constant factors may show up.

Reply:
> `Seq` isn't strict either, Seq's spine specifically isn't strict

Whoops, thanks for reminding me